trie/zk_merkle_state_tree.go

```diff
@@ -0,0 +1,120 @@
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/ethdb"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/trie/zk"
+)
+
+type ZkMerkleStateTree struct {
+	*ZkMerkleTree
+	preimage *preimageStore
+}
+
+func NewZkMerkleStateTree(rootHash common.Hash, db *Database) (*ZkMerkleStateTree, error) {
+	tree, err := zk.NewMerkleTreeFromHash(zkt.NewHashFromBytes(rootHash.Bytes()), db.Get)
+	if err != nil {
+		return nil, err
+	}
+	return &ZkMerkleStateTree{ZkMerkleTree: NewZkMerkleTree(tree, db), preimage: db.preimages}, nil
+}
+
+func NewEmptyZkMerkleTree(db *Database) *ZkMerkleStateTree {
+	return &ZkMerkleStateTree{ZkMerkleTree: NewZkMerkleTree(zk.NewEmptyMerkleTree(), db), preimage: db.preimages}
+}
+
+func (z *ZkMerkleStateTree) GetKey(kHashBytes []byte) []byte {
+	// TODO: use a kv cache in memory
+	k, err := zkt.NewBigIntFromHashBytes(kHashBytes)
+	if err != nil {
+		log.Error("Unhandled trie error in ZkMerkleStateTree.GetKey", "err", err)
+	}
+	if z.db.preimages == nil {
+		return nil
+	}
+	return z.db.preimages.preimage(common.BytesToHash(k.Bytes()))
+}
+
+func (z *ZkMerkleStateTree) GetStorage(_ common.Address, key []byte) ([]byte, error) {
+	return z.get(key)
+}
+
+func (z *ZkMerkleStateTree) GetAccount(address common.Address) (*types.StateAccount, error) {
+	if blob, err := z.get(address[:]); blob == nil || err != nil {
+		return nil, err
+	} else {
+		return types.UnmarshalStateAccount(blob)
+	}
+}
+
+func (z *ZkMerkleStateTree) get(key []byte) ([]byte, error) {
+	if hash, err := z.hash(key); err != nil {
+		return nil, err
+	} else {
+		return z.MerkleTree.Get(hash[:])
+	}
+}
+
+func (z *ZkMerkleStateTree) UpdateStorage(_ common.Address, key, value []byte) error {
+	return z.update(key, value)
+}
+
+func (z *ZkMerkleStateTree) UpdateAccount(address common.Address, account *types.StateAccount) error {
+	values, _ := account.MarshalFields()
+	var v []byte
+	for _, value := range values {
+		v = append(v, value.Bytes()...)
+	}
+	return z.update(address[:], v)
+}
+
+func (z *ZkMerkleStateTree) UpdateContractCode(_ common.Address, _ common.Hash, _ []byte) error {
+	return nil
+}
+
+func (z *ZkMerkleStateTree) update(key, value []byte) error {
+	if hash, err := z.hash(key); err != nil {
+		return err
+	} else {
+		return z.MerkleTree.Update(hash[:], value)
+	}
+}
+
+func (z *ZkMerkleStateTree) DeleteStorage(_ common.Address, key []byte) error { return z.delete(key) }
+func (z *ZkMerkleStateTree) DeleteAccount(address common.Address) error       { return z.delete(address[:]) }
+func (z *ZkMerkleStateTree) delete(key []byte) error {
+	if hash, err := z.hash(key); err != nil {
+		return err
+	} else {
+		return z.MerkleTree.Delete(hash[:])
+	}
+}
+
+func (z *ZkMerkleStateTree) Prove(key []byte, proofDb ethdb.KeyValueWriter) error {
+	return z.prove(key, proofDb, func(node zk.TreeNode) error {
+		value := node.CanonicalValue()
+		if leaf, ok := node.(*zk.LeafNode); ok {
```

Comment:
If it's not a LeafNode, why not return err?

Reply:
Both zktrie and mpt don't return errors just because they can't find leaf nodes. I fit that spec